File: apps/daemon/src/routes/integrations.ts
import type {
  DaemonConfig
} from "@orquester/config";
import type { IntegrationsResponse, UpdateIntegrationsRequest } from "@orquester/api";
import type { FastifyInstance } from "fastify";
import { writeFile } from "node:fs/promises";
import { getIntegrationAvailability } from "../integrations/catalog";

interface IntegrationRoutesOptions {
  config: DaemonConfig;
  configPath: string;
  local: boolean;
  apply: (integrations: Record<string, boolean>) => void;
}

export function registerIntegrationRoutes(app: FastifyInstance, options: IntegrationRoutesOptions): void {
  app.get("/api/integrations", async (): Promise<IntegrationsResponse> => {
    const available = await getIntegrationAvailability();
    return {
      integrations: available.map((integration) => ({
        ...integration,
        enabled: Boolean(options.config.integrations[integration.id]) && integration.available
      }))
    };
  });

  app.put<{ Body: Partial<UpdateIntegrationsRequest> }>(
    "/api/integrations",
    async (request, reply): Promise<IntegrationsResponse | void> => {
      if (!options.local) {
        return reply.code(403).send({ code: "FORBIDDEN", message: "Integrations can only be changed locally." });
      }
      const requested = request.body?.integrations;
      if (!requested || typeof requested !== "object") {
        return reply.code(400).send({ code: "INVALID_REQUEST", message: "integrations object required." });
      }
      const available = await getIntegrationAvailability();
      const known = new Set(available.map((integration) => integration.id));
      const integrations = { ...options.config.integrations };
      for (const [id, enabled] of Object.entries(requested)) {
        if (known.has(id) && typeof enabled === "boolean") integrations[id] = enabled;
      }
      await writeFile(options.configPath, `${JSON.stringify({ ...options.config, integrations }, null, 2)}\n`, "utf8");
      options.config.integrations = integrations;
      options.apply(integrations);
      return {
        integrations: available.map((integration) => ({
          ...integration,
          enabled: Boolean(integrations[integration.id]) && integration.available
        }))
      };
    }
  );
}

File: apps/worker/src/routes/integrations.rs
//! `/api/integrations`.

use crate::api_types::{ApiError, IntegrationsResponse, UpdateIntegrationsRequest};
use crate::state::{AppState, TransportMode};
use axum::extract::State;
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;

fn err(status: StatusCode, code: &str, message: impl Into<String>) -> Response {
    (status, Json(ApiError::new(code, message))).into_response()
}

async fn build_response(state: &AppState) -> IntegrationsResponse {
    let daemon = state.services.config.daemon.read().await;
    let available = crate::catalog::integration_availability().await;
    let integrations = available
        .into_iter()
        .map(|mut entry| {
            entry.enabled = daemon.integrations.get(&entry.id).copied().unwrap_or(false) && entry.available;
            entry
        })
        .collect();
    IntegrationsResponse { integrations }
}

pub async fn list(State(state): State<AppState>) -> Response {
    Json(build_response(&state).await).into_response()
}

pub async fn update(State(state): State<AppState>, Json(body): Json<UpdateIntegrationsRequest>) -> Response {
    if state.options.mode != TransportMode::Local {
        return err(StatusCode::FORBIDDEN, "FORBIDDEN", "Integrations can only be changed locally.");
    }
    let available = crate::catalog::integration_availability().await;
    let known: std::collections::HashSet<String> = available.iter().map(|entry| entry.id.clone()).collect();

    {
        let mut daemon = state.services.config.daemon.write().await;
        for (id, enabled) in &body.integrations {
            if known.contains(id) {
                daemon.integrations.insert(id.clone(), *enabled);
            }
        }
        let config_path = state.services.config.resolved.read().await.config_path.clone();
        let _ = crate::bootstrap::persist_config(&config_path, &daemon).await;
    }

    state.services.apply_integrations().await;
    Json(build_response(&state).await).into_response()
}
